src/Datadog.Trace/TraceContext.cs

```diff
@@ -1,53 +1,102 @@
 using System;
 using System.Collections.Generic;
-using System.Diagnostics;
-using System.Linq;
 using Datadog.Trace.Logging;
 
 namespace Datadog.Trace
 {
     internal class TraceContext : ITraceContext
     {
-        private static ILog _log = LogProvider.For<TraceContext>();
+        private static readonly ILog Log = LogProvider.For<TraceContext>();
 
-        private object _lock = new object();
-        private IDatadogTracer _tracer;
-        private List<Span> _spans = new List<Span>();
-        private int _openSpans = 0;
-        private DateTimeOffset _start;
-        private Stopwatch _sw;
+        private readonly object _lock = new object();
+        private readonly List<Span> _spans = new List<Span>();
+
+        private Span _rootSpan;
+        private int _openSpans;
+        private SamplingPriority? _samplingPriority;
+        private bool _samplingPriorityLocked;
 
         public TraceContext(IDatadogTracer tracer)
         {
-            _tracer = tracer;
-            _start = DateTimeOffset.UtcNow;
-            _sw = Stopwatch.StartNew();
+            Tracer = tracer;
         }
 
-        public DateTimeOffset UtcNow()
+        public DateTimeOffset UtcNow => DateTimeOffset.UtcNow;
+
+        public IDatadogTracer Tracer { get; }
+
+        /// <summary>
+        /// Gets or sets sampling priority set by user code.
+        /// Once the sampling priority is locked, setting this is a no-op.
+        /// </summary>
+        public SamplingPriority? SamplingPriority
         {
-            return _start.Add(_sw.Elapsed);
+            get => _samplingPriority;
+            set
+            {
+                if (!_samplingPriorityLocked)
+                {
+                    _samplingPriority = value;
+                }
+            }
         }
 
         public void AddSpan(Span span)
         {
             lock (_lock)
             {
+                if (_openSpans == 0)
+                {
+                    _rootSpan = span;
```

Comment:
This is where we'll want to do:

```c#
_rootSpan = span;

if (_rootSpan.SamplingPriority == null)
{
    // Determine an initial sampling priority for this trace, but don't lock it yet
    _rootSpan.SamplingPriority = Tracer.Sampler.GetSamplingPriority();
}
```

Reply:
We want to determine all our priority up front so that we can distribute to other services.

Reply:
Sorry, my code example here isn't actually useful, since we don't have `_rootSpan.SamplingPriority`, I meant `if (SamplingPriority == null) { SamplingPriority = Tracer.Sampler.GetSamplingPriority(); }`

Reply:
The idea here is when we initial create a root span we determine the starting sampling priority if one hasn't been set yet.

Then, at any point during that root span's lifecycle we might update the sampling priority, either having a user manually change it, or we inherit from parsing distributed tracing headers.

This will save us from having to have "LockSamplingPriority" being called at a certain time to determine an initial value, it'll really just be needed to stop the value from being changed.

Reply:
Thanks! This makes more sense and make the locking logic simpler, since it doesn't need to set sampling priority anymore. f0d4634baebb941533a36e796db8402996083515